torch/csrc/cudnn/Conv.cpp

```diff
@@ -84,77 +84,139 @@ BenchmarkCache<cudnnConvolutionBwdDataAlgo_t> bwd_data_algos;
-{
-  cudnnConvolutionBwdFilterAlgo_t algo;
-  if (benchmark) {
-    if (bwd_filter_algos.find(conv.params, algo)) {
-      return algo;
-    }
+  static void getAlgorithm(cudnnHandle_t handle, const Convolution& conv, cudnnConvolutionBwdDataAlgo_t* algo) {
+    CHECK(cudnnGetConvolutionBackwardDataAlgorithm(handle, conv.wdesc.desc,
+        conv.odesc.desc, conv.cdesc.desc, conv.idesc.desc,
+        CUDNN_CONVOLUTION_BWD_DATA_PREFER_FASTEST, 0, algo));
+  }
+
+  static void getWorkspaceSize(cudnnHandle_t handle, const Convolution& conv, cudnnConvolutionBwdDataAlgo_t algo, size_t* workspaceSize) {
+    CHECK(cudnnGetConvolutionBackwardDataWorkspaceSize(handle, conv.wdesc.desc,
+        conv.odesc.desc, conv.cdesc.desc, conv.idesc.desc, algo,
+        workspaceSize));
+  }
+};
+
+template<>
+struct algorithm_search<cudnnConvolutionBwdFilterAlgo_t> {
+  static constexpr auto DEFAULT_ALGO = CUDNN_CONVOLUTION_BWD_FILTER_ALGO_1;
+  static BenchmarkCache<cudnnConvolutionBwdFilterAlgo_t>& cache() {
+    return bwd_filter_algos;
+  }
+
+  static cudnnConvolutionBwdFilterAlgoPerf_t findAlgorithm(cudnnHandle_t handle, const Convolution& conv) {
     int algoCount;
     cudnnConvolutionBwdFilterAlgoPerf_t perfResults;
     CHECK(cudnnFindConvolutionBackwardFilterAlgorithm(handle, conv.idesc.desc,
         conv.odesc.desc, conv.cdesc.desc, conv.wdesc.desc, 1, &algoCount, &perfResults));
-    bwd_filter_algos.insert(conv.params, perfResults.algo);
-    return perfResults.algo;
+    return perfResults;
+  }
+
+  static void getAlgorithm(cudnnHandle_t handle, const Convolution& conv, cudnnConvolutionBwdFilterAlgo_t* algo) {
+    CHECK(cudnnGetConvolutionBackwardFilterAlgorithm(handle, conv.idesc.desc,
+        conv.odesc.desc, conv.cdesc.desc, conv.wdesc.desc,
+        CUDNN_CONVOLUTION_BWD_FILTER_PREFER_FASTEST, 0, algo));
+  }
+
+  static void getWorkspaceSize(cudnnHandle_t handle, const Convolution& conv, cudnnConvolutionBwdFilterAlgo_t algo, size_t* workspaceSize) {
+    CHECK(cudnnGetConvolutionBackwardFilterWorkspaceSize(handle, conv.idesc.desc,
+        conv.odesc.desc, conv.cdesc.desc, conv.wdesc.desc, algo, workspaceSize));
+  }
+};
+
+template<typename algo_t>
+Workspace chooseAlgorithm(
+    THCState* state, cudnnHandle_t handle, const Convolution& conv,
+    bool benchmark, algo_t* algo)
+{
+  using search = algorithm_search<algo_t>;
+  if (benchmark) {
+    auto& cache = search::cache();
+    if (!cache.find(conv.params, algo)) {
+      auto perfResults = search::findAlgorithm(handle, conv);
+      if (perfResults.status == CUDNN_STATUS_SUCCESS) {
+        *algo = perfResults.algo;
+      } else {
+        *algo = search::DEFAULT_ALGO;
+      }
+      cache.insert(conv.params, *algo);
+    }
+  } else {
+    search::getAlgorithm(handle, conv, algo);
+  }
+
+  size_t workspace_size;
+  search::getWorkspaceSize(handle, conv, *algo, &workspace_size);
+  try {
+    return Workspace(state, workspace_size);
```

Comment:
That sounds pretty awful for those who'll be running with tight memory constraints and won't be using `benchmark` 😕